Catalogue item change form

Macroz opened the pull request:
This implements a first (maybe rough) version of #837 and closes it.

The looping is done in the UI so that if there is a problem, it's stopped, but the user can opt to rerun the loop again. Also the new table select with filter options can be used creatively to select the forms to change e.g. by filtering to existing use of some form.

There can be some confusion for users which functions are "bulk edits" and which affect only one item. However re-rendering the buttons on the table is too slow and "jumpy". We could perhaps have an area for only bulk operations, but that we can decide based on feedback. At least it is possible to go to view an item and back, and not lose the selection.

Since this is a "power tool", we should spend extra time making sure this is properly tested and does not cause loss of work / ops work.

![Screenshot_2019-11-25_13-01-49](https://user-images.githubusercontent.com/823661/69535294-466efd00-0f84-11ea-9f8b-7beb1a98b8b1.png)
![Screenshot_2019-11-25_13-02-16](https://user-images.githubusercontent.com/823661/69535302-4969ed80-0f84-11ea-9823-6166d028d4db.png)
![Screenshot_2019-11-25_13-02-31](https://user-images.githubusercontent.com/823661/69535304-4bcc4780-0f84-11ea-83bd-2ff9dab86722.png)

# Definition of Done / Review checklist

## Reviewability
- [x] link to issue
- ~[ ] note if PR is on top of other PR~
- ~[ ] note if related change in rems-deploy repo~
- [x] consider adding screenshots for ease of review

## API
- [x] API is documented and shows up in Swagger UI
- [x] API is backwards compatible or completely new
- ~[ ] Events are backwards compatible~

## Documentation
- [x] update changelog if necessary
- [x] add or update docstrings for namespaces and functions
- [x] components are added to guide page
- ~[ ] documentation _at least_ for config options (i.e. docs folder)~
- ~[ ] ADR for major architectural decisions or experiments~

## ~Different installations~

## Testing
- [x] complex logic is unit tested
- [x] valuable features are integration / browser / acceptance tested automatically

## ~Accessibility~
Note, the accessibility is based on existing table, dropdown and table selection features so there is no change.

## Follow-up
- ~[ ] new tasks are created for pending or remaining tasks~
- [x] no critical TODOs left to implement

luontola commented:
I think the selected items should be reset when entering the catalogue page. 

Also, seem that you can select items from multiple non-overlapping searches, which might be ok as a design decision (e.g. if it's not easy to find all items to change with a single search), but it becomes confusing when they are still selected after visiting other pages. Resetting the selection on entering the catalogue page should fix that.

luontola commented on resources/translations/en.edn:
```diff
@@ -185,6 +188,7 @@
              :show "Show"
              :show-less "Show less"
              :show-more "Show more"}
+  :change-form {:form-selection "Form"}
```
Should more translation keys be in this map? At least the intro text.

Macroz replied:
I don't see the need for premature duplication of the translation texts. We don't intend to normalize / dedupe them anyway for other things.

luontola replied:
I mean that `:t.administration/change-form-intro` is not a generic text used on many administration page, but specific to the change-form page, so a renaming it to `:t.change-form/intro` would be more cohesive.

luontola replied:
Likewise `:t.administration/change-form` could be renamed to `:t.change-form/title`

Macroz replied:
I will move the intro, that text is actually leftover from earlier time when it was still on the same page. I think we often use the more general texts in the specific sub-pages than vice versa so I think I will leave the latter to admin.

luontola commented on src/clj/rems/api/catalogue_items.clj:
```diff
@@ -48,7 +48,7 @@
 (s/defschema ChangeFormResponse
   {:success s/Bool
-   :new-catalogue-item-id s/Int})
+   :catalogue-item-id s/Int})
```
This breaking API change should be mentioned in the changelog.

Macroz replied:
It's not a breaking API change when this is the PR where the API is introduced in.

luontola replied:
Ok.

luontola commented on src/cljs/rems/administration/catalogue_items.cljs:
```diff
@@ -60,12 +61,30 @@
+   {:disabled (when (empty? items) :disabled)
+    :on-click (fn []
+                (rf/dispatch [:rems.administration.change-catalogue-item-form/enter-page items])
+                (navigate! "/administration/catalogue-items/change-form") )}
```
Why is enter-page dispatched here in the link instead of the router, like all other pages?

luontola replied:
If it's because of the items parameter, that could also be transmitted as query parameters in the URL.

Macroz replied:
We **don't** want to reset everything on page enter i.e. when you change URL. That would disallow view of single item and back while having something selected.

Macroz replied:
Query parameters would introduce a lot of duplication or coupling between the table where everything is selected and change form page.

luontola commented on src/cljs/rems/administration/catalogue_items.cljs:
```diff
@@ -104,8 +123,8 @@
                       :sort-value (if checked? 1 2)})
-           :commands {:td [:td.commands
-                           [to-catalogue-item (:id item)]
+           :commands {:td [:td.commands {:on-click #(.stopPropagation %)}
```
Why stopPropagation? A clarifying comment would be necessary.

Macroz replied:
Otherwise clicking on a button link in commands will propagate to row selection.

luontola replied:
Please add a comment which says that.

luontola replied:
Clicking the resource/form/workflow links also selects the row, so each of those would also need the same treatment. Or then in the click handler for the row check what was the target element of the click event, and only select the row if it was a `td` element. Or then just rely on the checkboxes for row selection.

Macroz replied:
I think I'll try improving the row click handler. That way the row selection click handling does not propagate to table users.

Macroz replied:
The click handler now checks that the target was a `td`, `th` or `tr` element which I think is good enough for this purpose as the default action.

luontola commented on src/cljs/rems/administration/catalogue_items.cljs:
```diff
@@ -131,11 +150,16 @@
      [table/table catalogue-table]]))
 
+(defn- items-by-id [items ids]
+  (filter (comp ids :id) items))
```
For me items-by-id would be the name for a map of format `{id item}`, but this function is something else. Maybe extract the predicate to a named function and use filter directly.

Macroz replied:
I think it's missing an s.

luontola replied:
Something like `filter-items-by-ids` or `select-items-with-ids` would make it more obvious what the function does.

okahilak replied:
I have to say that I agree with Esko here. :) With its current name, I was expecting the return value to be a map with ids as keys.

luontola commented on src/cljs/rems/administration/catalogue_items.cljs:
```diff
@@ -144,7 +168,10 @@
-            [to-create-catalogue-item]
-            [status-flags/display-archived-toggle #(rf/dispatch [::fetch-catalogue])]
+            [:div.commands.text-left.pl-0
+             [ create-catalogue-item-button]
```
Extra space.

Macroz replied:
> I think the selected items should be reset when entering the catalogue page.
> 
> Also, seem that you can select items from multiple non-overlapping searches, which might be ok as a design decision (e.g. if it's not easy to find all items to change with a single search), but it becomes confusing when they are still selected after visiting other pages. Resetting the selection on entering the catalogue page should fix that.

It would prevent the user from viewing specific rows while trying to select the right ones because the list resets on every navigation. Like you can read from commit messages I chose to explicitly allow multiple filters as a feature and disallow archive toggle here.

Macroz commented:
> I think the selected items should be reset when entering the catalogue page.
> 
> Also, seem that you can select items from multiple non-overlapping searches, which might be ok as a design decision (e.g. if it's not easy to find all items to change with a single search), but it becomes confusing when they are still selected after visiting other pages. Resetting the selection on entering the catalogue page should fix that.

It would prevent the user from viewing specific rows while trying to select the right ones because the list resets on every navigation. Like you can read from commit messages I chose to explicitly allow multiple filters as a feature and disallow archive toggle here.

luontola commented on src/cljs/rems/administration/change_catalogue_item_form.cljs:
```diff
@@ -0,0 +1,140 @@
+          (flash-message/show-default-success! :top [text :t.administration/change-form])
+
+          (not= (:formid item) (:form/id form))
+          (rf/dispatch [::change-catalogue-item-form! (:id item) form (partial form-change-loop (rest items) form)])
```
`(partial ` could be just `#(`, especially since all the arguments are already defined here, so it's not the typical case of partial application.

Macroz replied:
That is not the meaning of partial to me but I'll do the change.

luontola commented on src/cljs/rems/administration/change_catalogue_item_form.cljs:
```diff
@@ -0,0 +1,140 @@
+          :else (recur (rest items) form))))
+
+(defn- all-items-have-the-form-already? [items form]
+  (every? (comp #{(:form/id form)} :formid) items))
```
I'm having a hard time understanding this line of code, since it uses a set as a function together with comp, both of which are complicated enough to understand alone.

luontola replied:
If I understand correctly, this could be refactored to `#(= (:form/id form) (:formid %))`

Macroz replied:
Using a set for this is idiomatic usage in Clojure.  I think using `comp` is also idiomatic here. However the alternative is also pretty clear so I'll change it.

luontola commented on src/cljs/rems/administration/change_catalogue_item_form.cljs:
```diff
@@ -0,0 +1,140 @@
+                    :td [:td.form
+                         [atoms/link nil
+                          (str "/administration/forms/" (:formid item))
+                          value]]})})
```
A better name for the `value` variable would be `form-name`.

luontola commented on src/cljs/rems/table.cljs:
```diff
@@ -259,9 +267,63 @@
+ (fn [db [_ table rows]]
+   (let [new-db (assoc-in db [::selected-rows (:id table)] (set (map :key rows)))]
+     (when-let [on-select (:on-select table)]
+       (on-select (get-in new-db [::selected-rows (:id table)])))
```
Instead of looking up the selected rows from `new-db` right after putting them there, they could be saved to a local variable for reuse: `(let [selected-rows (set (map :key rows))]`

Macroz replied:
Yes, maybe not worth the change but I did it anyway.

okahilak commented on resources/translations/en.edn:
```diff
@@ -185,6 +187,8 @@
              :show "Show"
              :show-less "Show less"
              :show-more "Show more"}
+  :change-form {:change-form-intro "You are about to change the form of the following catalogue items. The changes will be made one at a time, so that each old item is terminated and a new copy with the chosen form will be activated."
```
This is referred to as :t.administration/change-form-intro in the code, instead of :t.change-form/change-form-intro, as it's defined here.

okahilak commented on resources/translations/fi.edn:
```diff
@@ -183,6 +185,8 @@
              :show "Näytä"
              :show-less "Näytä vähemmän"
              :show-more "Näytä lisää"}
+  :change-form {:change-form-intro "Olet muuttamassa seuraavien aineistojen lomaketta. Muutokset tehdään yksi kerrallaan siten, että vanha aineisto lopetetaan ja uuteen aktivoitavaan kopioon laitetaan valitsemasi lomake."
```
Could it say "poistetaan käytöstä" instead of "lopetetaan"? I think it would be more aligned with the other terminology.

okahilak replied:
Also, how about "uuteen aktivoitavaan kopioon laitetaan valitsemasi lomake" -> "uusi aktivoitava kopio käyttää valitsemaasi lomaketta"? I think using "laittaa" in the above meaning still has a somewhat casual style, so it's not very formal language.

okahilak commented on src/cljs/rems/administration/catalogue_items.cljs:
```diff
@@ -60,12 +61,30 @@
 
-(defn- to-create-catalogue-item []
+(rf/reg-event-db
+ ::set-selected-items!
```
nit: I checked the codebase for other similar cases and it looks like it's not customary in our codebase to use exclamation mark with these reg-event-db setters, so you might want to align this with that convention.

Macroz replied:
Makes sense

okahilak commented:
Hmm, not sure if this is intended or a bug, but I couldn't figure out the logic behind this behavior:

1. First, no items are selected:

![image](https://user-images.githubusercontent.com/49641103/69710844-c3c37a80-1108-11ea-8c40-85783c36ee0c.png)

2. After I select one resource:

![image](https://user-images.githubusercontent.com/49641103/69710874-d0e06980-1108-11ea-8d9f-1fcd8c9cb654.png)

3. After I write something to the search field:

![image](https://user-images.githubusercontent.com/49641103/69710899-da69d180-1108-11ea-99a3-f5c4f0ce64ce.png)

4. After I press the "minus", it looks like this:

![image](https://user-images.githubusercontent.com/49641103/69710934-e3f33980-1108-11ea-8aa9-62dcf6d24d3d.png)

5. After I press the "check" repeatedly, it stays looking like this:

![image](https://user-images.githubusercontent.com/49641103/69710977-f40b1900-1108-11ea-92b5-6fda95058d16.png)

okahilak commented:
Also, a simpler case that seems to work unintuitively:

1. An empty search:

![image](https://user-images.githubusercontent.com/49641103/69712106-f2425500-110a-11ea-81d5-305d1d1c91c9.png)

2. After pressing the checkbox:

![image](https://user-images.githubusercontent.com/49641103/69712145-fff7da80-110a-11ea-89ad-0bf9c3176b5b.png)

3. Repeatedly pressing the checkbox keeps it checked:

![image](https://user-images.githubusercontent.com/49641103/69712159-071ee880-110b-11ea-85d5-c4c5941bbf7f.png)

4. Emptying the search field removes the tick mark:

![image](https://user-images.githubusercontent.com/49641103/69712195-19008b80-110b-11ea-9e93-7c12f0ca2fca.png)

5. Writing something to the search field that empties the search results adds the tick mark back:

![image](https://user-images.githubusercontent.com/49641103/69712326-4d744780-110b-11ea-9f95-20a5e9b4a228.png)

okahilak commented on src/cljs/rems/administration/change_catalogue_item_form.cljs:
```diff
@@ -0,0 +1,140 @@
+
+
+(rf/reg-event-fx
+ ::change-catalogue-item-form!
```
Also here, it seems that the current convention in the code is to not have the exclamation mark in the event name, even when the effect calls a clojure function with side effects (such as post! here).

okahilak commented:
Two more comments here:

- I'd like to see "Back" button on "Change Form" page, for consistency with other pages, and to get more easily back to "Catalogue" page (it took me one or two seconds to figure out that I have to press "Catalogue" again on the tab listing to get back to Catalogue page => a bit too long).

- When having several archived Catalogue items after changing forms and I sort by resource name, the table looks like this:

![image](https://user-images.githubusercontent.com/49641103/69720966-7e10ad00-111c-11ea-84f8-a41421a8be1c.png)

You might want to consider having active catalogue item first on that list when there's a tie. It's kind of minor, not necessarily directly related to this patch, but I don't see anyone doing it later on either if it's not done now.

Before this patch, there were more rarely ties between entries, so subtleties of sorting mattered less.

Macroz commented:
> Two more comments here:
> 
> * I'd like to see "Back" button on "Change Form" page, for consistency with other pages, and to get more easily back to "Catalogue" page (it took me one or two seconds to figure out that I have to press "Catalogue" again on the tab listing to get back to Catalogue page => a bit too long).
> * When having several archived Catalogue items after changing forms and I sort by resource name, the table looks like this:
> 
> ![image](https://user-images.githubusercontent.com/49641103/69720966-7e10ad00-111c-11ea-84f8-a41421a8be1c.png)

Added the button.

> 
> You might want to consider having active catalogue item first on that list when there's a tie. It's kind of minor, not necessarily directly related to this patch, but I don't see anyone doing it later on either if it's not done now.
> 
> Before this patch, there were more rarely ties between entries, so subtleties of sorting mattered less.

I added a tie breaker by reverse sorting by created so that latest is first. That is essentially the same as active but handles the history better.

luontola commented on src/cljs/rems/administration/catalogue_items.cljs:
```diff
@@ -78,7 +98,9 @@
-           :name {:value (get-localized-title item language)}
+           :name {:value (get-localized-title item language)
+                  :sort-value [(get-localized-title item language)
+                               (- (time-coerce/to-long (:start item)))]} ; secondary sort by created, reverse
```
This secondary sort doesn't work when sorting by resource, form or workflow.

Macroz replied:
Secondary support was not built to the table and this is only to improve the functionality of the default sort in this (arguably common) case. I'm not going to implement secondary sort to the table here.

luontola commented:
The colspan for the "show all" row is off by one. It's visible when hovering the row with the mouse.

<img width="1245" alt="Screenshot 2019-12-03 12 16 42" src="https://user-images.githubusercontent.com/42678/70042232-2e702c80-15c7-11ea-8cc1-bb7b932149d2.png">

okahilak commented:
The patch is fine by me now, thanks!
